src/modules/mc_pos_control/PositionControl/PositionControl.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+void PositionControl::setInputSetpoint(const vehicle_local_position_setpoint_s &setpoint)
+{
+	_pos_sp = Vector3f(setpoint.x, setpoint.y, setpoint.z);
+	_vel_sp = Vector3f(setpoint.vx, setpoint.vy, setpoint.vz);
+	_acc_sp = Vector3f(setpoint.acceleration);
+	_thr_sp = Vector3f(setpoint.thrust);
+	_yaw_sp = setpoint.yaw;
+	_yawspeed_sp = setpoint.yawspeed;
+}
+
+void PositionControl::setConstraints(const vehicle_constraints_s &constraints)
+{
+	_constraints = constraints;
+
+	// For safety check if adjustable constraints are below global constraints. If they are not stricter than global
+	// constraints, then just use global constraints for the limits.
+	if (!PX4_ISFINITE(constraints.tilt) || (constraints.tilt > _lim_tilt)) {
+		_constraints.tilt = _lim_tilt;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_up) || (constraints.speed_up > _lim_vel_up)) {
+		_constraints.speed_up = _lim_vel_up;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_down) || (constraints.speed_down > _lim_vel_down)) {
+		_constraints.speed_down = _lim_vel_down;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_xy) || (constraints.speed_xy > _lim_vel_horizontal)) {
+		_constraints.speed_xy = _lim_vel_horizontal;
+	}
+}
+
+bool PositionControl::update(const float dt)
+{
+	_positionControl();
+	_velocityControl(dt);
+
+	_yawspeed_sp = PX4_ISFINITE(_yawspeed_sp) ? _yawspeed_sp : 0.f;
+	_yaw_sp = PX4_ISFINITE(_yaw_sp) ? _yaw_sp : _yaw;
+
+	const bool valid_acceleration_setpoint = PX4_ISFINITE(_acc_sp(0))
+			&& PX4_ISFINITE(_acc_sp(1))
+			&& PX4_ISFINITE(_acc_sp(2));
+	return valid_acceleration_setpoint;
+}
+
+void PositionControl::_positionControl()
+{
+	Vector3f vel_sp_position = (_pos_sp - _pos).emult(_gain_pos_p);
+	_addIfNotNanVector(_vel_sp, vel_sp_position);
+	_addIfNotNanVector(vel_sp_position, Vector3f());
+
+	// Constrain horizontal velocity by prioritizing the velocity component along the
+	// the desired position setpoint over the feed-forward term.
+	const Vector2f vel_sp_xy = ControlMath::constrainXY(Vector2f(vel_sp_position),
+				   Vector2f(_vel_sp - vel_sp_position), _constraints.speed_xy);
+	_vel_sp(0) = vel_sp_xy(0);
+	_vel_sp(1) = vel_sp_xy(1);
+	// Constrain velocity in z-direction.
+	_vel_sp(2) = math::constrain(_vel_sp(2), -_constraints.speed_up, _constraints.speed_down);
+}
+
+void PositionControl::_velocityControl(const float dt)
+{
+	const float hover_scale = CONSTANTS_ONE_G / _hover_thrust;
+
+	// PID velocity control
+	Vector3f vel_error = _vel_sp - _vel;
+	Vector3f acc_sp_velocity = vel_error.emult(_gain_vel_p) + _vel_int + _vel_dot.emult(_gain_vel_d);
+
+	// For backwards compatibility of the gains
+	acc_sp_velocity *= hover_scale;
+	// No control input from setpoints or corresponding states which are NAN
+	_addIfNotNanVector(_acc_sp, acc_sp_velocity);
+
+	_accelerationControl();
+
+	// Apply Anti-Windup in vertical direction
+	if ((_thr_sp(2) >= -_lim_thr_min && vel_error(2) >= 0.0f) ||
+	    (_thr_sp(2) <= -_lim_thr_max && vel_error(2) <= 0.0f)) {
```

Comment:
I liked the intermediate variables that were before: `u_max = -_lim_thr_min` to then check `if (_thr_sp(2) >= u_max)`

Reply:
That comment made me rethink how it's implemented and you're right it's confusing to check `_thr_sp(2) >= -_lim_thr_min` and it's only necessary because the parameters are in scalar values and all the vectors including `_thr_sp` are in NED frame. And such cases are exactly what setters are for so I fixed it by converting from [0,1] to [-1,0] in the setter and commented there so it's all more clear.  Fixed in https://github.com/PX4/Firmware/pull/12072/commits/3ecdfbe3775d870c976b0953bee75c99e7a07d72